### PR TITLE
add function to throw current weapon

### PR DESCRIPTION
**Pull request by commy2**
**When merged this pull request will:**
- title

```sqf
player call ACE_hitreactions_fnc_throwWeapon
```
Let it fly.

### CODE REVIEW

File: addons/hitreactions/functions/fnc_throwWeapon.sqf [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#define THROW_TORQUE 0.2
+
+params ["_unit"];
+private _weapon = currentWeapon _unit;
+if (_weapon isEqualTo "") exitWith {objNull};
+
+private _data = weaponsItems _unit select {_x select 0 == _weapon} select 0;
+
+private _holder = createVehicle ["WeaponHolderSimulated", [0,0,0], [], 0, "CAN_COLLIDE"];
+_holder addWeaponWithAttachmentsCargoGlobal [_data, 1];
```
Comment (dedmen) [inline addons/hitreactions/functions/fnc_throwWeapon.sqf]: :smiley:

Comment (commy2) [reply addons/hitreactions/functions/fnc_throwWeapon.sqf]: It's a really fucking great command, dude. Look at L25. That's all you need to get the relevant info. Splendid job.

Comment (dedmen) [reply addons/hitreactions/functions/fnc_throwWeapon.sqf]: Well my command draft would've been slightly different format than weaponsItems, but whoever copy-pasted it fixed that, dunno who to thank for that.

Comment (commy2) [reply addons/hitreactions/functions/fnc_throwWeapon.sqf]: That's not something you admit.

File: addons/hitreactions/functions/fnc_throwWeapon.sqf [hunk 2]
```diff
@@ -0,0 +1,42 @@
+#include "script_component.hpp"
+/*
+ * Author: commy2
+ * Makes the unit throw their currently selected weapon. Unit must be local.
+ *
+ * Arguments:
+ * 0: unit <OBJECT>
```
Comment (jonpas) [inline addons/hitreactions/functions/fnc_throwWeapon.sqf]:
```suggestion
 * 0: Unit <OBJECT>
```

File: addons/hitreactions/functions/fnc_throwWeapon.sqf [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ * Return Value:
+ * Weapon Holder
+ *
+ * Example:
+ * player call ACE_hitreactions_fnc_throwWeapon
```
Comment (jonpas) [inline addons/hitreactions/functions/fnc_throwWeapon.sqf]:
```suggestion
 * player call ace_hitreactions_fnc_throwWeapon
```

File: addons/hitreactions/functions/fnc_throwWeapon.sqf [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ * Public: No
+ */
+
+#define THROW_VELOCITY 1.5
+#define THROW_TORQUE 0.2
+
+params ["_unit"];
```
Comment (jonpas) [inline addons/hitreactions/functions/fnc_throwWeapon.sqf]:
```suggestion
params ["_unit"];

```

File: addons/hitreactions/functions/fnc_throwWeapon.sqf [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+private _data = weaponsItems _unit select {_x select 0 == _weapon} select 0;
+
+private _holder = createVehicle ["WeaponHolderSimulated", [0,0,0], [], 0, "CAN_COLLIDE"];
+_holder addWeaponWithAttachmentsCargoGlobal [_data, 1];
+
+private _vDir = _unit weaponDirection _weapon;
+private _vLat = vectorNormalized (_vDir vectorCrossProduct [0,0,1]);
+private _vUp = _vLat vectorCrossProduct _vDir;
+
+private _position = _unit modelToWorldWorld (_unit selectionPosition "RightHand") vectorAdd (_vLat vectorMultiply 0.59);
```
Comment (jonpas) [inline addons/hitreactions/functions/fnc_throwWeapon.sqf]: `0.59` magic number?

File: addons/hitreactions/functions/fnc_throwWeapon.sqf [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+private _weapon = currentWeapon _unit;
+if (_weapon isEqualTo "") exitWith {objNull};
+
+private _data = weaponsItems _unit select {_x select 0 == _weapon} select 0;
+
+private _holder = createVehicle ["WeaponHolderSimulated", [0,0,0], [], 0, "CAN_COLLIDE"];
+_holder addWeaponWithAttachmentsCargoGlobal [_data, 1];
+
+private _vDir = _unit weaponDirection _weapon;
+private _vLat = vectorNormalized (_vDir vectorCrossProduct [0,0,1]);
```
Comment (jonpas) [inline addons/hitreactions/functions/fnc_throwWeapon.sqf]:
```suggestion
private _vLat = vectorNormalized (_vDir vectorCrossProduct [0, 0, 1]);
```

File: addons/hitreactions/functions/fnc_throwWeapon.sqf [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+#define THROW_VELOCITY 1.5
+#define THROW_TORQUE 0.2
+
+params ["_unit"];
+private _weapon = currentWeapon _unit;
+if (_weapon isEqualTo "") exitWith {objNull};
+
+private _data = weaponsItems _unit select {_x select 0 == _weapon} select 0;
+
+private _holder = createVehicle ["WeaponHolderSimulated", [0,0,0], [], 0, "CAN_COLLIDE"];
```
Comment (jonpas) [inline addons/hitreactions/functions/fnc_throwWeapon.sqf]:
```suggestion
private _holder = createVehicle ["WeaponHolderSimulated", [0, 0, 0], [], 0, "CAN_COLLIDE"];
```

File: addons/hitreactions/functions/fnc_throwWeapon.sqf (hunk 5 above)
Comment (commy2) [reply addons/hitreactions/functions/fnc_throwWeapon.sqf]: Yeah, where the proxy is (after rotation) in the GWH dummy model.

File: addons/hitreactions/functions/fnc_throwWeapon.sqf (hunk 6 above)
Comment (commy2) [reply addons/hitreactions/functions/fnc_throwWeapon.sqf]: Not how we do arrays with only numbers not variables, but I guess it is not consistant.

**Comment by commy2** [comment]
Open question if this should be included
- [ ] "ace_weaponThrown" event [_unit, _weaponHolder, _data<weapon/attachments/mags>]
Either local or global event.

**Comment by ColdEvul** [comment]
Can this some how be added to units surrendering moduel?
Whould be awesome :)

**Comment by commy2** [comment]
Sure. That's pretty simple after this PR is merged:
```sqf
["ace_captiveStatusChanged", {
    params ["_unit", "_state", "_type"];

    if (local _unit && {_state && {_type == "SetSurrendered"}}) then {
        _unit call ace_hitreactions_fnc_throwWeapon;
    };
}] call CBA_fnc_addEventHandler;
```
Can also be put into the mission, e.g. the init.sqf.

**Comment by LorenLuke** [comment]
Can this be used to throw satchel charges from ace explosives, or the like?

**Comment by commy2** [comment]
No, this throws a weapon, including binocular. Does not work for magazines/ammo of the Put weapon.

**Comment by dedmen** [comment]
> Can this be used to throw satchel charges from ace explosives, or the like?

somewhat, you'd have to create a weapon holder to throw it (which might not show the actual explosive model, but instead the item model which might just be a small bag), and then when the holder landed on the ground and became static, delete it and spawn the real ammo, which will look terrible.

**Comment by commy2** [comment]
Ok, done here.

**Comment by commy2** [comment]
MERGE?

File: addons/hitreactions/functions/fnc_throwWeapon.sqf (hunk 6 above)
Comment (jonpas) [reply addons/hitreactions/functions/fnc_throwWeapon.sqf]: Where does it say it's not how we do it? Space after a comma.

**Comment by commy2** [comment]
@jonpas It doesn't say this anywhere, but that's how I've been doing it. But if you insists.

File: addons/hitreactions/functions/fnc_throwWeapon.sqf (hunk 5 above)
Comment (jonpas) [reply addons/hitreactions/functions/fnc_throwWeapon.sqf]: Macro though?

**Comment by commy2** [comment]
As you wish.

**Comment by commy2** [comment]
The throw angle (inclination on top of the weapon direction) is currently hard coded to atan(1/2)=+26.6 degree. I suppose it could be rewritten such that the angle can be defined as macro.

Does this sound good? Implementation is trivial thanks to CBA_fnc_rotateVector3D.

**Comment by jonpas** [comment]
Hmm, doesn't sound worth it, just a self-explanatory macro as you did is fine.

**Comment by commy2** [comment]
Atm it is realized by: `_vDir vectorAdd (_vUp vectorMultiply 2)`, so
1\*e_dir + 2\*e_up
dir being the adjacent and up the opposite

**Comment by commy2** [comment]
Derp, it is as simple as this.

**Comment by jonpas** [comment]
Nice! MERGE! :joy:

**Pull request merged**